crates/ide-completion/src/render/function.rs

```diff
@@ -278,31 +278,44 @@ pub(super) fn add_call_parens<'b>(
         (snippet, "(…)")
     };
     if ret_type.is_unit() && ctx.config.add_semicolon_to_unit {
-        let next_non_trivia_token =
-            std::iter::successors(ctx.token.next_token(), |it| it.next_token())
-                .find(|it| !it.kind().is_trivia());
-        let in_match_arm = ctx.token.parent_ancestors().try_for_each(|ancestor| {
-            if ast::MatchArm::can_cast(ancestor.kind()) {
-                ControlFlow::Break(true)
-            } else if matches!(ancestor.kind(), SyntaxKind::EXPR_STMT | SyntaxKind::BLOCK_EXPR) {
-                ControlFlow::Break(false)
-            } else {
-                ControlFlow::Continue(())
+        let inside_closure_ret = ctx.token.parent_ancestors().try_for_each(|ancestor| {
+            match_ast! {
+                match ancestor {
+                    ast::BlockExpr(_) => ControlFlow::Break(false),
+                    ast::ClosureExpr(_) => ControlFlow::Break(true),
+                    _ => ControlFlow::Continue(())
+                }
             }
         });
```

Comment:
Kind of missed this last time but we should do this traversal in the completion analysis and store the result where relevant, otherwise we redo this work for every function completion item